standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/MetaStoreDirectSql.java

```diff
@@ -515,6 +529,803 @@ public List<String> getMaterializedViewsForRewriting(String dbName) throws MetaE
+          params[paramIndex++] = partPrivilege.getPrivilege();
+        }
+        try (QueryWrapper query = new QueryWrapper(pm.newQuery("javax.jdo.query.SQL", batchQueryText))) {
+          executeWithArray(query.getInnerQuery(), params, batchQueryText);
+        }
+      }
+    };
+    insertInBatch(PART_PRIVS, columns, columnCount, row, rowCount, bec);
+  }
+
+  private void insertPartitionColPrivilegeInBatch(Map<Long, MPartitionColumnPrivilege> partColumnGrantIdToPrivilege,
+      Map<Long, Long> partColumnGrantIdToPartId) throws MetaException {
+    int rowCount = partColumnGrantIdToPrivilege.size();
+    String columns = "(\"PART_COLUMN_GRANT_ID\",\"AUTHORIZER\",\"COLUMN_NAME\",\"CREATE_TIME\",\"GRANT_OPTION\","
+        + "\"GRANTOR\",\"GRANTOR_TYPE\",\"PART_ID\",\"PRINCIPAL_NAME\",\"PRINCIPAL_TYPE\",\"PART_COL_PRIV\")";
+    String row = "(?,?,?,?,?,?,?,?,?,?,?)";
+    int columnCount = 11;
+    BatchExecutionContext bec = new BatchExecutionContext() {
+      final Iterator<Map.Entry<Long, MPartitionColumnPrivilege>> it
+          = partColumnGrantIdToPrivilege.entrySet().iterator();
+      @Override
+      public void execute(String batchQueryText, int batchRowCount, int batchParamCount) throws MetaException {
+        Object[] params = new Object[batchParamCount];
+        int paramIndex = 0;
+        for (int index = 0; index < batchRowCount; index++) {
+          Map.Entry<Long, MPartitionColumnPrivilege> entry = it.next();
+          MPartitionColumnPrivilege partColumnPrivilege = entry.getValue();
+          params[paramIndex++] = entry.getKey();
+          params[paramIndex++] = partColumnPrivilege.getAuthorizer();
+          params[paramIndex++] = partColumnPrivilege.getColumnName();
+          params[paramIndex++] = partColumnPrivilege.getCreateTime();
+          params[paramIndex++] = partColumnPrivilege.getGrantOption();
+          params[paramIndex++] = partColumnPrivilege.getGrantor();
+          params[paramIndex++] = partColumnPrivilege.getGrantorType();
+          params[paramIndex++] = partColumnGrantIdToPartId.get(entry.getKey());
+          params[paramIndex++] = partColumnPrivilege.getPrincipalName();
+          params[paramIndex++] = partColumnPrivilege.getPrincipalType();
+          params[paramIndex++] = partColumnPrivilege.getPrivilege();
+        }
+        try (QueryWrapper query = new QueryWrapper(pm.newQuery("javax.jdo.query.SQL", batchQueryText))) {
+          executeWithArray(query.getInnerQuery(), params, batchQueryText);
+        }
+      }
+    };
+    insertInBatch(PART_COL_PRIVS, columns, columnCount, row, rowCount, bec);
+  }
+
+  /**
+   * Add partitions in batch using direct SQL
+   * @param parts list of partitions
+   * @param partPrivilegesList list of partition privileges
+   * @param partColPrivilegesList list of partition column privileges
+   * @throws MetaException
+   */
+  public void addPartitions(List<MPartition> parts, List<List<MPartitionPrivilege>> partPrivilegesList,
```

Comment:
nit: seems better to create a new class for adding the partitions by direct sql. For example, there is a class named `DirectSqlUpdateStat` for inserting column statistics.

Reply:
agreed